File: IDP - Dynamic Mapping/Assets/Scripts/GeometryClustering/CircleBuilder.cs
using System.Collections.Generic;
using UnityEngine;

public class CircleBuilder {

    // Points representing this circle:
    private List<Point> points;

    // Regression parameters of the circle (Rx, Ry):
    private Vector2 Rxy = new Vector2();

    // Current center of the circle:
    private Vector2 center;

    public CircleBuilder() {
        points = new List<Point>();
    }

    public CircleBuilder(List<Point> points) {
        this.points = points;

        foreach (Point point in points)
            Rxy += point.position;

        center = Rxy / points.Count;
    }

    public void AddPoint(Point point) {
        points.Add(point);

        Rxy += point.position;

        center = Rxy / points.Count;
    }

    public Circle Build() {
        // Compute the radius of the circle:
        float R = ComputeCircleRadius();

        // Build the circle:
        Circle circle = new Circle(center.x, center.y, R);

        // Match all the points to the built circle:
        foreach (Point point in points) point.MatchToPrimitive(circle);

        return circle;
    }

    private float ComputeCircleRadius() {
        int n = points.Count;

        if (n == 1)
            return 0;

        float sigma_x2 = 0, sigma_y2 = 0;
        foreach (Point p in points) {
            sigma_x2 += (p.position.x - center.x) * (p.position.x - center.x);
            sigma_y2 += (p.position.y - center.y) * (p.position.y - center.y);
        }

        // The division by 'n-1' instead of 'n' is known as Bessel's correction,
        // which corrects the bias in the estimation of the points variance:
        sigma_x2 /= n - 1; sigma_y2 /= n - 1;

        return Mathf.Sqrt(sigma_x2 + sigma_y2);
    }

    // Return the distance between the circle center and the given point:
    public float DistanceFrom(Point point) {
        return (point.position - center).magnitude;
    }

    public int PointsCount() {
        return points.Count;
    }
}

File: DynamicMapping/mapping/geometry_maps/circle_builder.py
import numpy as np

class CircleBuilder:

    def __init__(self, points : dict | None = None):
        # Regression parameters of the circle [Rx, Ry]:
        self.Rxy = np.zeros(2)

        # Points representing this circle:
        if points is None:
            self.points = {}
            self.points['positions'] = []
            self.points['covariances'] = []
        
        else:
            self.points = points

            for position in points['positions']:
                self.Rxy += position

        self.center = self.Rxy / len(self.points['positions'])

    def add_point(self, position, covariance):        
        self.points['positions'].append(position)
        self.points['covariances'].append(covariance)
        
        self.Rxy += position

        self.center = self.Rxy / len(self.points['positions'])

    def build(self):
        # Compute the radius of the circle:
        R = self.compute_circle_radius()

        # Build the circle:
        # circle = Circle(center.x, center.y, R);
        circle = (self.center, R)

        # Match all the points to the built circle:
        # foreach (Point point in points) point.MatchToPrimitive(circle);

        return circle

    def compute_circle_radius(self):
        n = len(self.points['positions'])

        if (n == 1):
            return 0

        sigma = np.zeros(2)
        for position in self.points['positions']:
            sigma += (position - self.center) ** 2
        
        # The division by 'n-1' instead of 'n' is known as Bessel's correction,
        # which corrects the bias in the estimation of the points variance:
        return np.sqrt(np.sum(sigma / (n - 1)))

    def distance_from(self, point):
        """ Returns the distance between the circle center and the given point """
        return np.sqrt(np.sum((point[1] - self.center) ** 2))

    def points_count(self):
        return len(self.points['positions'])